src/libstd/sys/unix/fs.rs

```diff
@@ -463,18 +468,24 @@ pub fn rmdir(p: &Path) -> io::Result<()> {
 pub fn readlink(p: &Path) -> io::Result<PathBuf> {
     let c_path = try!(cstr(p));
     let p = c_path.as_ptr();
-    let mut len = unsafe { libc::pathconf(p as *mut _, libc::_PC_NAME_MAX) };
-    if len < 0 {
-        len = 1024; // FIXME: read PATH_MAX from C ffi?
-    }
-    let mut buf: Vec<u8> = Vec::with_capacity(len as usize);
-    unsafe {
-        let n = try!(cvt({
-            libc::readlink(p, buf.as_ptr() as *mut c_char, len as size_t)
-        }));
-        buf.set_len(n as usize);
+
+    let mut buf = Vec::with_capacity(256);
+
+    loop {
+        let buf_read = try!(cvt(unsafe { libc::readlink(
+                                                    p,
+                                                    buf.as_mut_ptr() as *mut libc::c_char,
+                                                    buf.capacity() as libc::size_t) })) as usize;
+        if buf_read != buf.capacity() {
+            unsafe { buf.set_len(buf_read); }
+            buf.shrink_to_fit();
+
+            return Ok(PathBuf::from(OsString::from_vec(buf)));
+        }
+
+        let new_cap = buf.capacity() * 2;
+        buf.reserve(new_cap);
```

Comment:
Ah yeah I guess that makes sense. In any case, however, these should all follow the same strategy instead of reinventing the wheel. Let's have them all take the `reserve(1)` route to punt to `Vec` as much as possible for figuring out its reallocation strategy